intake/cli/server/server.py

```diff
@@ -107,15 +112,33 @@ def get(self):
             else:
                 start = int(page_offset)
                 stop = int(page_offset) + int(page_size)
-            page = itertools.islice(cat.walk(depth=1).items(), start, stop)
+            if query_list:
+                # This could be a search of a search of a serach (etc.).
+                # Progressively apply each search and then page through the
+                # final results.
+                for args, kwargs in query_list:
```

Comment:
I wonder if, for the mongo case, there is something to be done to merge the queries/dicts rather than chaining them in this loop. Forgive me if we already talked about the possibility

Reply:
Yeah, in the PR description (a long time ago now haha!) I pitched:

> This PR implements progressive search by caching the chain of queries accumulated over chained searched and iteratively applying them each time. This could be made more efficient: one could imagine adding an optional `optimized_multi_search()` method that expects a list of queries and does internal optimization. I'd be happy to include that in this PR, but I'll hold off until we agree that the basic outline is good.

In the Mongo case, that would be:

```py
def optimized_compound_search(self, *queries):
    return self.serach({'$and': queries})
```

I think it should be a separate method so that we don't have to alter the signature of `search(...)` to accommodate it, which I think would compromise the usability of `search(...)`. I am imagining that the server would check for this method, use it if it's implemented and fall back on the iterative approach if it is not. What do you think?

Reply:
I don't think another route or method is a good thing, I agree with you. Consolidation of query terms may be done by the can on the server if it supports it, and the client should know nothing about it.

Still I have something I think is a misunderstanding - maybe we can talk it through live.

Let cat be a local reference to a bluesky catalog on a server, backed by mongo.

The user performs `cat2 = cat.search(...)`. The arguments are passed to the server, it executes the search and stores a new cat instance in the cache, returning an ID to the client. cat2 now has one entry in its query list, and a remote ID to pull entries from.

The user performs `cat3 = cat2.search(...)`. The arguments passed to the server include both the ID of the cat in the server cache and two elements of query, the previous one and the one added in this step. On the server side, which of the following occurs:
- use ID to fetch cat2 from the cache, and do cat2.search(query_list[-1]); noting that cat2 remembers the query that created it
- do cat.search(query_list[0]).search(query_list[1])
- do cat.search(query list), which is the equivalent of the previous option, but with the possibility of an optimised query in the future.

Reply:
The concern I have with that approach is that it sometimes passes `search(one_query)` and other times passes `search(query_list)`, which means `search` has to be smart enough to distinguish between them, and in general it might not be clear. Moreover, the "query" might not just one item; it could involve the use of kwargs like `depth`. That is the problem I am aiming to solve with a separate, dev-oriented method that explicitly takes a list of `(args, kwargs)` pairs.